domain/cloud/state/state.go

```diff
@@ -0,0 +1,441 @@
+		var (
+			cloudUUID string
+			cert      string
+		)
+		if err := rows.Scan(&cloudUUID, &cert); err != nil {
+			return nil, errors.Trace(err)
+		}
+		_, ok := result[cloudUUID]
+		if !ok {
+			result[cloudUUID] = []string{}
+		}
+		result[cloudUUID] = append(result[cloudUUID], cert)
+	}
+	return result, rows.Err()
+}
+
+// loadRegions loads the regions for the specified clouds, returning
+// a map of results keyed on cloud uuid.
+func loadRegions(tx *sql.Tx, cloudUUIDS []string) (map[string][]cloud.Region, error) {
+	q := fmt.Sprintf(`
+		SELECT
+			cloud_uuid, name, endpoint, identity_endpoint, storage_endpoint
+		FROM cloud_region
+		WHERE cloud_uuid IN (%s)`, database.SliceToPlaceholder(cloudUUIDS))[1:]
+
+	rows, err := tx.Query(q, transform.Slice(cloudUUIDS, func(s string) any { return s })...)
+	if err != nil && err != sql.ErrNoRows {
+		return nil, errors.Trace(err)
+	}
+	defer func() { _ = rows.Close() }()
+
+	var result = map[string][]cloud.Region{}
+	for rows.Next() {
+		var (
+			cloudUUID string
+			dbRegion  CloudRegion
+		)
+		if err := rows.Scan(&cloudUUID, &dbRegion.Name, &dbRegion.Endpoint, &dbRegion.IdentityEndpoint, &dbRegion.StorageEndpoint); err != nil {
+			return nil, errors.Trace(err)
+		}
+		_, ok := result[cloudUUID]
+		if !ok {
+			result[cloudUUID] = []cloud.Region{}
+		}
+		result[cloudUUID] = append(result[cloudUUID], cloud.Region{
+			Name:             dbRegion.Name,
+			Endpoint:         dbRegion.Endpoint,
+			IdentityEndpoint: dbRegion.IdentityEndpoint,
+			StorageEndpoint:  dbRegion.StorageEndpoint,
+		})
+	}
+	return result, rows.Err()
+}
+
+// UpsertCloud inserts or updates the specified cloud.
+func (st *State) UpsertCloud(ctx context.Context, cloud cloud.Cloud) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		// Get the cloud UUID - either existing or make a new one.
+		var cloudUUID string
+		row := tx.QueryRow("SELECT uuid FROM cloud WHERE name = ?", cloud.Name)
+		err := row.Scan(&cloudUUID)
+		if err != nil && err != sql.ErrNoRows {
+			return errors.Trace(err)
+		}
+		if err != nil {
+			cloudUUID = utils.MustNewUUID().String()
+		}
+
+		if err := upsertCloud(ctx, tx, cloudUUID, cloud); err != nil {
+			return errors.Annotate(err, "updating cloud")
+		}
+
+		if err := updateAuthTypes(ctx, tx, cloudUUID, cloud.AuthTypes); err != nil {
+			return errors.Annotate(err, "updating cloud auth types")
+		}
+
+		if err := updateCACerts(ctx, tx, cloudUUID, cloud.CACertificates); err != nil {
+			return errors.Annotate(err, "updating cloud CA certs")
+		}
+
+		if err := updateRegions(ctx, tx, cloudUUID, cloud.Regions); err != nil {
+			return errors.Annotate(err, "updating cloud regions")
+		}
+
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func upsertCloud(ctx context.Context, tx *sql.Tx, cloudUUID string, cloud cloud.Cloud) error {
+	dbCloud, err := dbCloudFromCloud(tx, cloudUUID, cloud)
```

Comment:
Do a join in the main query. See the lease store for an example.

Reply:
If you are talking about what is done in ClaimLease() then that doesn't appear to correctly return an error if the lease type is invalid, like we do here if a cloud with an unknown type is saved.